utils/unittest/epoch_builder.go

```diff
@@ -395,14 +396,13 @@ func (builder *EpochBuilder) AddBlocksWithSeals(n int, counter uint64) *EpochBui
 		// Given the last 2 blocks in state A <- B when we add block C it will contain the following.
 		// - seal for A
 		// - execution result for B
-		a := builder.blocks[len(builder.blocks)-2]
 		b := builder.blocks[len(builder.blocks)-1]
 
 		receiptB := ReceiptForBlockFixture(b)
 
 		block := BlockWithParentFixture(b.Header)
 		seal := Seal.Fixture(
-			Seal.WithResult(a.Payload.Results[0]),
+			Seal.WithResult(b.Payload.Results[0]),
```

Comment:
This was a bug, we don't need last 2 blocks, because the result in the last block is already pointing at the last second block. We should always seal the result in the last block.

Reply:
The result is in the previous block's payload, the seal is in the current block's payload. The seal and receipt are not in the same block.

If we seal the result 2 blocks away, we are sealing a result that is already being sealed. 

This wasn't caught until I added the index when a block is sealed in this PR, the error says a block is sealed twice, after debugging I found this bug here.

Reply:
It seems we create put a receipt and seal in the same payload, but the receipt's result is in the previous block. 

Why not putting receipt and result into the same block? The previous blocks were like this. For instance[ block G's payload has result and receipt for block F](https://github.com/onflow/flow-go/blob/3104f4a591072c9ed5a43fbd5767a1466ee89a93/utils/unittest/epoch_builder.go#L297-L298)

Reply:
## ⚠️ 

> we cannot include a receipt and seal for the same block

Correct; including a seal in the same block that also includes the result is _impossible_. This is because:
*  We first need to publish a block incorporating the result. 
* Then, we need a child block in addition for the source of randomness.
* The source of randomness of the block that incorporates the result defines the verification node assignment

Hence, the block that incorporates the result must be published, otherwise, verifiers don't know which chunks to check.

https://www.notion.so/dapperlabs/How-ConsensusNode-choose-seals-and-receipts-to-include-in-the-next-block-payload-72c178bccd4a416d8cd4380af4cfb6be

Reply:
I think my fix is correct here. 

What we have is a chain with each block contains a result for previous and a seal for previous of previous block:
```
A <- B (result_A) <- C (result_B, seal_A) <- D(resultC, seal_B)
```

And we want to add similar blocks like this. 

But the tests was creating a wrong seal. It is supposed to seal its grandparent block (height - 2), instead it seals the result in its grandparent block, which is its grand-grandparent block (height - 3), causing the grand-grantparent to be sealed again, which is causing the tests to fail due to inserting the seal index twice.

My fix is to say seal the result in parent block, which is the result for its grandparent block.

Reply:
Thanks for the example, that really helped to understand why your fix is correct. 👍